test/server/listener_manager_impl_test.cc

```diff
@@ -2761,53 +2788,48 @@ TEST_F(ListenerManagerImplWithRealFiltersTest, TransparentListenerEnabled) {
-  NiceMock<Api::MockOsSysCalls> os_sys_calls;
-  TestThreadsafeSingletonInjector<Api::OsSysCallsImpl> os_calls(&os_sys_calls);
-
-  const std::string yaml = TestEnvironment::substitute(R"EOF(
+  const envoy::api::v2::Listener listener = parseListenerFromV2Yaml(R"EOF(
     name: FreebindListener
     address:
       socket_address: { address: 127.0.0.1, port_value: 1111 }
     filter_chains:
     - filters:
     freebind: true
-  )EOF",
-                                                       Network::Address::IpVersion::v4);
-  if (ENVOY_SOCKET_IP_FREEBIND.has_value()) {
-    EXPECT_CALL(listener_factory_, createListenSocket(_, _, _, true))
-        .WillOnce(
-            Invoke([this](Network::Address::InstanceConstSharedPtr, Network::Address::SocketType,
-                          const Network::Socket::OptionsSharedPtr& options,
-                          bool) -> Network::SocketSharedPtr {
-              EXPECT_NE(options.get(), nullptr);
-              EXPECT_EQ(options->size(), 1);
-              EXPECT_TRUE(
-                  Network::Socket::applyOptions(options, *listener_factory_.socket_,
-                                                envoy::api::v2::core::SocketOption::STATE_PREBIND));
-              return listener_factory_.socket_;
-            }));
-    EXPECT_CALL(os_sys_calls, setsockopt_(_, ENVOY_SOCKET_IP_FREEBIND.value().first,
-                                          ENVOY_SOCKET_IP_FREEBIND.value().second, _, sizeof(int)))
-        .WillOnce(Invoke([](int, int, int, const void* optval, socklen_t) -> int {
-          EXPECT_EQ(1, *static_cast<const int*>(optval));
-          return 0;
-        }));
-    manager_->addOrUpdateListener(parseListenerFromV2Yaml(yaml), "", true);
-    EXPECT_EQ(1U, manager_->listeners().size());
-  } else {
-    // MockListenerSocket is not a real socket, so this always fails in testing.
-    EXPECT_THROW_WITH_MESSAGE(
-        manager_->addOrUpdateListener(parseListenerFromV2Yaml(yaml), "", true), EnvoyException,
-        "MockListenerComponentFactory: Setting socket options failed");
-    EXPECT_EQ(0U, manager_->listeners().size());
-  }
+  )EOF");
+
+  endToEndSockoptTest(listener, envoy::api::v2::core::SocketOption::STATE_PREBIND,
+                      ENVOY_SOCKET_IP_FREEBIND,
+                      /* expected_value */ 1,
+                      /* expected_num_options */ 1);
+}
+
+// Validate that when tcp_fast_open_queue_length is set in the Listener, we see the socket option
+// propagated to setsockopt(). This is as close to an end-to-end test as we have
+// for this feature, due to the complexity of creating an integration test
+// involving the network stack. We only test the IPv4 case here, as the logic
+// around IPv4/IPv6 handling is tested generically in
+// socket_option_impl_test.cc.
+TEST_F(ListenerManagerImplWithRealFiltersTest, FastOpenListenerEnabled) {
+  const envoy::api::v2::Listener listener = parseListenerFromV2Yaml(R"EOF(
```

Comment:
I think the base config could also be moved to the test fixture and only the additional TCP option configured here, but up to you.